File: UserNotifications/Objective-C/NotificationService/NotificationService.m
//
//  NotificationService.m
//  NotificationService
//
//  Created by pro648 on 18/8/26
//  Copyright © 2018年 pro648. All rights reserved.
//

#import "NotificationService.h"

@interface NotificationService ()

@property (nonatomic, strong) void (^contentHandler)(UNNotificationContent *contentToDeliver);
@property (nonatomic, strong) UNMutableNotificationContent *bestAttemptContent;

@end

@implementation NotificationService

- (void)didReceiveNotificationRequest:(UNNotificationRequest *)request withContentHandler:(void (^)(UNNotificationContent * _Nonnull))contentHandler {
    self.contentHandler = contentHandler;
    self.bestAttemptContent = [request.content mutableCopy];
    
//    NSString *body = self.bestAttemptContent.body;
//    if (body.length > 0) {
//        self.bestAttemptContent.body = [NSString stringWithFormat:@"%@ by pro648",self.bestAttemptContent.body];
//    } else {
//        self.bestAttemptContent.body = @"Default Body by pro648";
//    }
//    self.contentHandler(self.bestAttemptContent);
    
    // Dig in the payload to get the attachment-url.
    if (! self.bestAttemptContent) {
        return;
    }
    
    NSString *attachmentURLAsString = (NSString *)request.content.userInfo[@"media-url"];
    if (!attachmentURLAsString) {
        return;
    }
    
    NSURL *attachmentURL = [NSURL URLWithString:attachmentURLAsString];
    
    // Download the image and pass it to attachments if not nil.
    [self downloadImageFromURL:attachmentURL completion:^(UNNotificationAttachment *attachment) {
            self.bestAttemptContent.attachments = [NSArray arrayWithObject:attachment];
            self.contentHandler(self.bestAttemptContent);
    }];
}

- (void)serviceExtensionTimeWillExpire {
    // Called just before the extension will be terminated by the system.
    // Use this as an opportunity to deliver your "best attempt" at modified content, otherwise the original push payload will be used.
    if (self.contentHandler && self.bestAttemptContent) {
        self.bestAttemptContent.title = @"Incoming Image";
        self.contentHandler(self.bestAttemptContent);
    }
}

- (void)downloadImageFromURL:(NSURL *)url completion:(void(^)(UNNotificationAttachment *attachment))completionHandler {
    NSURLSessionDownloadTask *task = [NSURLSession.sharedSession downloadTaskWithURL:url completionHandler:^(NSURL * _Nullable location, NSURLResponse * _Nullable response, NSError * _Nullable error) {
        // 1. Test URL and escape if URL not OK
        if (!location) {
            completionHandler(nil);
            return ;
        }
        
        // 2. Get current's user temporary directory path
        NSURL *urlPath = [NSURL fileURLWithPath:NSTemporaryDirectory()];
        // 3. Add proper ending to url path, in the case .jpg (The system validates the content of attached files before scheduling the corresponding notification request. If an attached file is corrupted, invalid, or of an unsupported file type, the notificaiton request is not scheduled for delivery.)
        NSString *uniqueURLString = [NSProcessInfo.processInfo.globallyUniqueString stringByAppendingString:@".png"];
        urlPath = [urlPath URLByAppendingPathComponent:uniqueURLString];
        
        // 4. Move downloadUrl to newly created urlPath
        if ([[NSFileManager defaultManager] moveItemAtURL:location toURL:urlPath error:NULL]) {
            UNNotificationAttachment *attachment = [UNNotificationAttachment attachmentWithIdentifier:@"picture" URL:urlPath options:nil error:nil];
            completionHandler(attachment);
        }else {
            completionHandler(nil);
        }
    }];
    [task resume];
}

@end

File: UserNotifications/Swift/NotificationService/NotificationService.swift
//
//  NotificationService.swift
//  NotificationService
//
//  Created by pro648 on 18/8/18
//  Copyright © 2018年 pro648. All rights reserved.
//

import UserNotifications

class NotificationService: UNNotificationServiceExtension {

    var contentHandler: ((UNNotificationContent) -> Void)?
    var bestAttemptContent: UNMutableNotificationContent?

    override func didReceive(_ request: UNNotificationRequest, withContentHandler contentHandler: @escaping (UNNotificationContent) -> Void) {
        self.contentHandler = contentHandler
        bestAttemptContent = (request.content.mutableCopy() as? UNMutableNotificationContent)
        
//        bestAttemptContent?.body = "\(bestAttemptContent?.body ?? "Default Body") by pro648"
//        contentHandler(bestAttemptContent!)
        
        // Dig in the payload to get the attachment-url.
        guard let bestAttemptContent = bestAttemptContent,
            let attachmentURLAsString = request.content.userInfo["media-url"] as? String,
            let attachmentURL = URL(string: attachmentURLAsString) else {
                return
        }
        
        // Download the image and pass it to attachments if not nil.
        downloadImageFrom(url: attachmentURL) {(attachment) in
            
            if attachment != nil {
                bestAttemptContent.attachments = [attachment!]
                contentHandler(bestAttemptContent)
            }
        }
    }
    
    override func serviceExtensionTimeWillExpire() {
        // Called just before the extension will be terminated by the system.
        if let contentHandler = contentHandler, let bestAttemptContent = bestAttemptContent {
            bestAttemptContent.title = "Incoming Image"
            contentHandler(bestAttemptContent)
        }
    }
}

extension NotificationService {
    private func downloadImageFrom(url: URL, with completionHandler: @escaping (UNNotificationAttachment?) -> Void) {
        let task = URLSession.shared.downloadTask(with: url) { (location, response, error) in
            // 1. Test URL and escape if URL not OK
            guard let location = location else {
                completionHandler(nil)
                return
            }
            
            // 2. Get current's user temporary directory path
            var urlPath = URL(fileURLWithPath: NSTemporaryDirectory())
            // 3. Add proper ending to url path, in the case .jpg (The system validates the content of attached files before scheduling the corresponding notification request. If an attached file is corrupted, invalid, or of an unsupported file type, the notificaiton request is not scheduled for delivery.)
            let uniqueURLString = ProcessInfo.processInfo.globallyUniqueString + ".png"
            urlPath = urlPath.appendingPathComponent(uniqueURLString)
            
            // 4. Move downloadUrl to newly created urlPath
            try? FileManager.default.moveItem(at: location, to: urlPath)
            
            // 5. Try adding the attachement and pass it to the completion handler
            do {
                let attachment = try UNNotificationAttachment(identifier: "picture", url: urlPath, options: nil)
                completionHandler(attachment)
            }
            catch {
                completionHandler(nil)
            }
        }
        task.resume()
    }
}
